### DISCUSSION TITLE
Simultaneous global optimization

### DISCUSSION THREAD

**Discussion by wcairnx** [JuliaQuantumControl/QuantumControl.jl#54] [Q&A]
Is it possible to optimize the shape of a single control waveform with multiple different overall amplitudes for multiple objectives? The idea is that I have a single fast control signal that can be multiplied by a different scalar amplitude for  each of wo qubits, and I would like to perform distinct gates on the two qubits. I would like to optimize both the control waveform shape and the two amplitudes to achieve robust but different gates on the two qubits.

**Comment by goerz** [comment]
That's an interesting problem! It's not something you can do *out of the box*, exactly, but QuantumControl.jl will certainly be helpful in this.

We don't have any global optimization methods. So you could write your own code or use some other package to create a swarm or grid for the two coupling parameters, and then use QuantumControl.jl with GRAPE or Krotov to optimize the amplitudes at each coupling parameter point. Then refine the swarm/grid based on for which points you had the most success in the amplitude optimization. That's pretty much exactly what we did in [Goerz *et al*. npj Quantum Information 3, 37 (2017)](https://dx.doi.org/10.1038/s41534-017-0036-0).

The alternative approach (although I haven't maybe fully thought this through) is to calculate a combined gradient for the amplitude values and the two coupling parameters. The calculation of gradients for parametrized pulses is still work-in-progress. We only recently defined the notion of a [`ParametrizedFunction`](https://juliaquantumcontrol.github.io/QuantumPropagators.jl/stable/api/quantumpropagators/#QuantumPropagators.Controls.ParameterizedFunction), of which a tunable coupling strength would be an example. Calculating gradients for such parameters would happen with the help of [QuantumGradientGenerator](https://github.com/JuliaQuantumControl/QuantumGradientGenerators.jl), along the lines of GOAT, see [Machnes *et al*. Phys. Rev. Lett. 120, 150401 (2018)](https://doi.org/10.1103/PhysRevLett.120.150401).

Once that's implemented, you could supposedly copy the amplitude gradient calculation from GRAPE and the parameter gradient calculation from GOAT and combine them into a single gradient vector and pass that to LBFGS.

But as I said, the parameter gradient calculation is still work-in-progress and even then, it would require a fair amount of problem-specific code to put everything together. So for the time being, I'd go with option 1, even though it's pretty costly in terms of numerical resources (hopefully, you have a cluster or a workstation with a significant number of cores)
